content/docs/fuzzing/c-cpp/12-libafl/index.md

```diff
@@ -0,0 +1,982 @@
+```
+
+The library crate should be compiled into a static library. To achieve this, we need to set the `crate-type` to `staticlib` in the `Cargo.toml` file:
+
+{{< customFigure "Final `Cargo.toml` file of the fuzzer" >}}
+```toml
+[package]
+name = "appsec_guide"
+version = "0.1.0"
+edition = "2021"
+
+[lib]
+crate-type = ["staticlib"]
+
+[dependencies]
+clap = { version = "4", features = ["derive"] }
+libafl = "0.13.1"
+libafl_bolts = "0.13.1"
+libafl_cc = "0.13.1"
+libafl_targets = { version = "0.13.1", features = ["libfuzzer", "sancov_pcguard_hitcounts"] }
+```
+{{< /customFigure >}}
+
+After setting up the basic project, clear the `src/lib.rs` file. The [full source code](https://github.com/trailofbits/testing-handbook/tree/main/materials/fuzzing/libafl/appsec_guide) for the following fuzzer can be found on GitHub.
+
+The following figure showcases a basic skeleton for a LibAFL fuzzer. The skeleton includes imports, a struct including parsed command line options, and the `libafl_main` function. Within the `libafl_main` function, we parse the command-line options, define the closure `run_client`, and finally set up everything using a LibAFL [Launcher](https://docs.rs/libafl/0.13.1/libafl/events/launcher/index.html). To set up the launcher instance, we define several options:
+
+* `.shmem_provider()` sets a shared-memory provider. Shared memory is used to synchronize fuzzing clients. The implementation is chosen based on the operating system.
+* `.configuration()` gives the fuzzer a name, which in this case is “default.”
+* `.monitor()` sets the monitor records aggregated statistics in a TOML file called `./fuzzer_stats.toml` and logs current statistics to the standard output.
+* `.run_client()` defines the lambda that creates the fuzzer and clients.
+* `.cores()` defines which and how many cores are used. Generally, you use the `Cores::from_cmdline` function to create a `Cores` instance from a string. For example, the string “0” creates a single client fuzzer that runs on the core with ID `0`. The string ”0,8-15” launches nine clients that run on core `0` and on the cores ranging from ID 8 until 15 (inclusive).
+* `.broker_port()` sets the port that the main broker process is using. Initial communication between a client and the main broker process happens over TCP.
+* `.remote_broker_addr()` can be used to set a broker process to connect to in case one already exists. Otherwise, this can be set to `None`.
+* `.stdout_file()` redirects the standard output of the SUT. You may want to set this to `/dev/null` if the output does not matter and the output is noisy. We comment this option here, as it may hide issues with the fuzzer because it also hides output printed in the `run_client` closure.
+
+{{< customFigure "Skeleton for a LibAFL fuzzer. [Full source code](https://github.com/trailofbits/testing-handbook/blob/main/materials/fuzzing/libafl/appsec_guide/src/lib.rs) can be found in the GitHub repository of the testing handbook." >}}
```

Comment:
of the Testing Handbook

https://github.com/trailofbits/testing-handbook?tab=readme-ov-file#writing-guidelines